ikra/soa/array_operator.h

```diff
@@ -0,0 +1,260 @@
+    STD_ARRAY result; \
+    for (size_t i = 0; i < ArraySize; ++i) { \
+      result[i] = left[i] symbol right[i]; \
+    } \
+    return result; \
+  } \
+  template<typename T, size_t N> \
+  void operator symbol ## = (std::array<T, N>& left, \
+                             std::array<T, N> right) { \
+    for (size_t i = 0; i < N; ++i) { \
+      left[i] symbol ## = right[i]; \
+    } \
+  } \
+  TEMPLATE_ARRAY_FIELD \
+  void operator symbol ## = (STD_ARRAY& left, \
+                             AOS_ARRAY_FIELD()& right) { \
+    for (size_t i = 0; i < ArraySize; ++i) { \
+      left[i] symbol ## = right[i]; \
+    } \
+  } \
+  TEMPLATE_ARRAY_FIELD \
+  void operator symbol ## = (STD_ARRAY& left, \
+                             SOA_ARRAY_FIELD()& right) { \
+    for (size_t i = 0; i < ArraySize; ++i) { \
+      left[i] symbol ## = right[i]; \
+    } \
+  }
+
+#define IKRA_DEFINE_STD_ARRAY_OPERATORS_M(symbol) \
+  template<typename T, size_t N> \
+  std::array<T, N> operator symbol (std::array<T, N> left, T right) { \
+    std::array<T, N> result; \
+    for (size_t i = 0; i < N; ++i) { \
+      result[i] = left[i] symbol right; \
+    } \
+    return result; \
+  } \
+  template<typename T, size_t N> \
+  std::array<T, N> operator symbol (T left, std::array<T, N> right) { \
+    std::array<T, N> result; \
+    for (size_t i = 0; i < N; ++i) { \
+      result[i] = left symbol right[i]; \
+    } \
+    return result; \
+  } \
+  template<typename T, size_t N> \
+  void operator symbol ## = (std::array<T, N>& left, T right) { \
+    for (size_t i = 0; i < N; ++i) { \
+      left[i] symbol ## = right; \
+    } \
+  }
+
+#define IKRA_DEFINE_AOS_ARRAY_FIELD_OPERATORS_A(symbol) \
+  TEMPLATE_ARRAY_FIELD \
+  STD_ARRAY operator symbol (AOS_ARRAY_FIELD()& left, \
+                             STD_ARRAY right) { \
+    return (STD_ARRAY&)left symbol right; \
+  } \
+  TEMPLATE_ARRAY_FIELD_W \
+  STD_ARRAY operator symbol (AOS_ARRAY_FIELD(1)& left, \
+                             AOS_ARRAY_FIELD(2)& right) { \
+    return (STD_ARRAY&)left symbol (STD_ARRAY&)right; \
+  } \
+  TEMPLATE_ARRAY_FIELD_W \
+  STD_ARRAY operator symbol (AOS_ARRAY_FIELD(1)& left, \
+                             SOA_ARRAY_FIELD(2)& right) { \
+    return (STD_ARRAY&)left symbol right; \
+  } \
+  TEMPLATE_ARRAY_FIELD \
+  void operator symbol ## = (AOS_ARRAY_FIELD()& left, \
+                             STD_ARRAY right) { \
+    (STD_ARRAY&)left symbol ## = right; \
+  } \
+  TEMPLATE_ARRAY_FIELD_W \
+  void operator symbol ## = (AOS_ARRAY_FIELD(1)& left, \
+                             AOS_ARRAY_FIELD(2)& right) { \
+    (STD_ARRAY&)left symbol ## = (STD_ARRAY&)right; \
+  } \
+  TEMPLATE_ARRAY_FIELD_W \
+  void operator symbol ## = (AOS_ARRAY_FIELD(1)& left, \
+                             SOA_ARRAY_FIELD(2)& right) { \
+    (STD_ARRAY&)left symbol ## = right; \
+  }
+
+#define IKRA_DEFINE_AOS_ARRAY_FIELD_OPERATORS_M(symbol) \
```

Comment:
Rename to `_SCALAR`. Provide both scalar and vector operations for all operators.